NIFI-12386 Adds processor FilterAttribute

EndzeitBegins opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary

Adds a new processor `FilterAttribute` to the `nifi-standard-processors` bundle.
See [NIFI-12386](https://issues.apache.org/jira/browse/NIFI-12386).

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

mosermw commented:
Wouldn't it be more intuitive to have this functionality in the UpdateAttribute processor, instead of creating a new processor?  You could create a "Retain Attributes Expression" or "Keep Attributes Expression" property on UpdateAttribute.  You would probably want to enforce setting either this new property or the existing "Delete Attributes Expression" property, not both.

I do think using a regular expression is a better approach than splitting a string using a delimiter.  Regular expressions support this already in the form ^(foo|bar|baz)$

EndzeitBegins commented:
@mosermw Thank you for your feedback. I've asked about the functionality on the community Slack 3 days ago and created the Jira issue (and the PR) based on the feedback I've got. 
The feedback, amongst others from @joewitt, proposed a new processor as part of the `nifi-standard-processors`. 

While I'm not completely against integrating the functionality into the `UpdateAttribute` processor, I like the idea of a separate processor, as it highlights the envisioned use case of regulating which attributes are available at a specified point in the flow.

mosermw commented:
Adding a new processor to the `nifi-standard-nar` with a clearly defined use case is certainly not going to be a bad thing.

I feel strongly that the attribute filter needs to be a regular expression, though.  If I want to retain 50 attributes out of 100 in a FlowFile, then I don't want to specify each one exactly.  A regex pattern, though a little more complex, is much more powerful and useful.  A regex can be written to be future-proof, to new attributes that get created, without having to update this processor's configuration every time I create new attributes in my flow.

EndzeitBegins commented:
@mosermw Thank for the review once more. 

I can see advantages to both approaches, explicit enumeration and pattern matching. 

While the first approach is more explicit and strict but more verbose, the later is more flexible and adaptive but less explicit and somewhat more complex.

What do you think about supporting both approaches? That is, introducing something like a `Attribute matching style` with allowed values `Enumeration`, `Regular Expression` (and possibly others such as glob pattern). 

When `Enumeration` is selected, the properties `Attributes to filter` and `Delimiter` are present. 
When `Regular expression` is selected on the other hand, these properties are not available but a property `Attribute filter regex` is present.

This way, both styles are supported, while keeping the properties available to the user at a time to a minimum.

What are your thoughts about this? If that's something we can agree on, I'm more than happy about suggestions for the property names.

mosermw commented:
@EndzeitBegins  I spent some time thinking about this.  I feel like using explicit enumeration of attributes would not be needed if operators could be trained to design a simple regex.  We could do this in the PropertyDescriptor description, even giving an example regex that acts like an enumeration of attributes.

However, regexes are hard for many people.  And I see you want to support Expression Language in the `Attributes to filter` filter .  I fully support doing this.  I can see use cases where using ExpLang to create a regex would not be ideal.  Possible, but not ideal.  So I would support an `Attribute Matching Strategy` PropertyDescriptor that lets you choose regex or enumeration to match attributes.

I'm not a fan, however, of a `Delimiter` PropertyDescriptor.  We aren't dealing with data that we can't control, the delimiter will only exist inside NiFi itself.  So, I feel like we can require a comma-separated enumeration here.  If operators have a pipe-separated enumeration already in an attribute, then ExpLang `replaceAll('|', ',')` can change the delimiter to a comma.

I do like the `Filter Mode` to make the processor remove or retain the matching attributes.  I would rename `Attributes to filter` to something like `Matching Attributes` or just `Attributes`.  If you do this, `Attributes` becomes the subject and `Filter Mode` becomes the action to take on that subject.

EndzeitBegins commented:
Thank you for your useful inputs @mosermw.

That sounds like a reasonable middle-ground. 

I'll have another go at the tests & implementation in the upcoming days, when I've time to spare. 
I'll mark the PR as draft again and get back to you, as soon as I incorporated the proposed changes.

mosermw reviewed (commented):
I took a quick glance at the code and noticed these.

EndzeitBegins commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttributes.java:
```diff
@@ -0,0 +1,216 @@
+    private static String getDelimiter(ProcessContext context, FlowFile flowFile) {
+        PropertyValue delimiterProperty = context.getProperty(DELIMITER);
+
+        if (flowFile != null) {
```
This code has been removed, as the delimiter is no longer based on a propety.

EndzeitBegins commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttributes.java:
```diff
@@ -0,0 +1,216 @@
+        final FilterMode filterMode = getFilterMode(context, flowFile);
+        final Set<String> attributesToRemove = computeAttributesToRemove(flowFile, filterMode, declaredAttributes);
+
+        session.removeAllAttributes(flowFile, attributesToRemove);
```
Changed.

EndzeitBegins commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttributes.java:
```diff
@@ -0,0 +1,216 @@
+    private static String getAttributeSet(ProcessContext context, FlowFile flowFile) {
+        PropertyValue attributeSetProperty = context.getProperty(ATTRIBUTE_SET);
+
+        if (flowFile != null) {
```
While we no longer pre-calculate / cache the attributeSet but a `Predicate<String>` instead, this is necessary as we can calculate it `@OnScheduled`, when no NiFi EL is used, or per FlowFile in the `onTrigger` whenever NiFi EL is used.
In the former case, the `flowFile` attribute is nullable.

Is this okay to you, or should this be refactored to make this more visible, e.g. using `Optional` or method overloads with and without `FlowFile flowFile` parameter?

EndzeitBegins commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttributes.java:
```diff
@@ -0,0 +1,216 @@
+        return properties;
+    }
+
+    private Set<String> preCalculatedAttributes;
```
Changed.

EndzeitBegins commented:
@Steigerwaldo and I adjusted to pull-request to support both enumeration and regex as matching strategy.
Additionally, the delimiter is no longer configurable as a property but statically defined. 

I would highly appreciate another code-review of the changes made, @mosermw.
I've added a remark regarding the null check.

EndzeitBegins commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttributes.java:
```diff
@@ -0,0 +1,275 @@
+                Specify the set of "Set of attributes to filter" using the delimiter comma ',', e.g. "my-property,other,filename".
+                """
+)
+public class FilterAttributes extends AbstractProcessor {
```
I changed the name in code, PR and Jira accordingly. Thanks for the hint. 👍🏻

exceptionfactory replied:
Thanks for making the adjustments!

EndzeitBegins commented:
Not quite sure why the CI pipeline failed. :( I didn't have any problems running locally with Linux / JDK 21.

At a first glance the error does not relate to any of the changes that are part of the PR. Maybe this is related to build problems reported by others on the dev mailing list which seemingly were introduced by #8053?
However, other pipeline runs seem to work just fine.

mosermw commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttribute.java:
```diff
@@ -0,0 +1,273 @@
+@SupportsBatching(defaultDuration = DefaultRunDuration.TWENTY_FIVE_MILLIS)
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@Tags({"attributes", "modification", "filter", "retain", "remove", "delete", "regex", "regular expression", "Attribute Expression Language"})
+@CapabilityDescription("Filters the Attributes of a FlowFile according to a specified strategy")
```
I think it would be nice to explain its capabilities more here.  Perhaps something like "Filters the attributes of a FlowFile by retaining specified attributes and removing the rest or by removing specified attributes and retaining the rest."

EndzeitBegins replied:
Well spotted, I adopted your suggestion.

mosermw commented on nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/FilterAttribute.java:
```diff
@@ -0,0 +1,273 @@
+            .description("Specifies the strategy to apply on filtered attributes. Either 'Remove' or 'Retain' only the matching attributes.")
+            .required(true)
+            .allowableValues(FILTER_MODE_VALUE_RETAIN, FILTER_MODE_VALUE_REMOVE)
+            .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
```
I understand the desire for FILTER_MODE to support expression language, but it presents some problems.
- the AllowableValue description can never be seen
- the operator has to type in either RETAIN or REMOVE, exactly
- if an ExpLang expression is entered, the processor is invalid because it doesn't match the allowableValues

If you remove the allowableValues restriction and replace it with a validator (see InvokeHTTP "HTTP Method"), then the processor can be valid with an ExpLang expression.  However, if that expression evaluates to an invalid FILTER_MODE, then the processor will admin yield the flowfile which stops data flowing through the processor.  You could catch this problem and send the flowfile to 'success' unmodified, but that hides the problem with the FlowFile's attributes.  You could have a failure relationship to deal with cases when the ExpLang expression causes a problem.

The easiest approach is to remove support for ExpLang for FILTER_MODE, though I don't have a guess for how many users would need this feature.

EndzeitBegins replied:
These are some good remarks, thank you.

I assume there won't be a lot of users that wish to use EL for the FILTER_MODE. I removed the support for now. In case using EL for this property is more desired than anticipated right now, it should be relatively easy to introduce proper support in one of the ways you mentioned.

mosermw reviewed (commented):
@EndzeitBegins I really love your code, you've done some things I have not seen before, so thank you for sharing.  I made just a couple of comments to ponder.

EndzeitBegins commented:
I made so small changes to the mentioned parts of the code.
I'll explain my thoughts on the changes in the corresponding review conversations.

Thank you for the reviews and the kind words @mosermw.

mosermw commented:
+1 this looks good.  Thanks for making the adjustments and for contributing this new processor!  Will merge to main.
